### PR TITLE
feat(cache): introduce cache component

### PR DESCRIPTION
**Pull request by azjezz**
Neu cache provides an atomic cache implementation, with multiple drivers allowing you to convert any existing PSR-6, PSR-16, Amphp, and Symfony cache object into a simple atomic cache.

---

TODO:

- [x] add tests
- [x] ~introduce redis driver based on amphp/redis~ not needed, amphp/redis provides amphp/cache implementation that could be used with the AmphpDriver
- [x] ~introduce filesystem driver based on amphp/file~ for another PR.
- [x] ~introduce mysql/pgsql drivers based on amphp/mysql and amphp/postgresql~ for another PR.

### CODE REVIEW

File: src/Cache/Cache.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+final class Cache implements CacheInterface
+{
+    /**
+     * @var Async\KeyedSemaphore<string, array{0: Closure(): mixed, 1: null|positive-int}, mixed>
+     */
+    private Async\KeyedSemaphore $semaphore;
+
+    public function __construct(private readonly Driver\DriverInterface $driver)
+    {
+        $this->semaphore = new Async\KeyedSemaphore(1, function (string $key, array $input): mixed {
```
Comment (azjezz) [inline src/Cache/Cache.php:19]: TODO: add KeyedSequence to PSL.

**Comment by azjezz** [comment]
static analysis is broken, will try to find a solution 🤔

File: src/Cache/Cache.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    /**
+     * @var Async\KeyedSequence<string, array{0: Closure(): mixed, 1: null|positive-int}, mixed>
+     */
+    private Async\KeyedSequence $sequence;
+
+    public function __construct(private readonly Driver\DriverInterface $driver)
+    {
+        $this->sequence = new Async\KeyedSequence(function (string $key, array $input): mixed {
+            try {
+                return $this->driver->get($key);
```
Comment (azjezz) [inline src/Cache/Cache.php:21]: while we do attempt a get() operation before waiting for semaphore in line 54, this is needed as the first get() might fail, while this succeed due to a pending compute operation that was finished at this step.

**Review by trowski — APPROVED** [review]
Seems a good start. We can address the issue of whether or not the cache interface should allow updating values at a later point.

**Pull request merged**